### PR TITLE
N2O Decomposition

### PR DESCRIPTION
**Pull request by whateverusername0**
## About the PR

Nitrous oxide now decomposes into oxygen and nitrogen when the mixture temperature is 1400K or above.

## Why / Balance

ss13 content

## Technical details

I've violently copied reactions code from /tg station, so it just works. 🤷‍♂

## Media

**The process**
https://github.com/space-wizards/space-station-14/assets/113782077/0c125217-2c7d-4206-a971-ca50074c3876
- [X] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

**Changelog**

:cl:
- add: Nitrous oxide now decomposes at high temperatures.

### CODE REVIEW

**Comment by Partmedia** [comment]
Thanks, awesome! I'll give this a test tomorrow my time :)

File: Content.Server/Atmos/Reactions/N2ODecompositionReaction.cs [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+    {
+        var cacheN2O = mixture.GetMoles(Gas.NitrousOxide);
+
+        var burnedFuel = cacheN2O * Atmospherics.N2ODecompositionRate * (mixture.Temperature - Atmospherics.N2ODecompositionMinScaleTemp) * (mixture.Temperature - Atmospherics.N2ODecompositionMaxScaleTemp) / Atmospherics.N2ODecompositionScaleDivisor;
+
+        //if (burnedFuel <= 0 || cacheN2O - burnedFuel < 0) return ReactionResult.NoReaction;
+
+        mixture.AdjustMoles(Gas.NitrousOxide, burnedFuel * Atmospherics.N2ODecompositionRate);
+        mixture.AdjustMoles(Gas.Nitrogen, -burnedFuel * Atmospherics.N2ODecompositionRate);
+        mixture.AdjustMoles(Gas.Oxygen, (-burnedFuel / 2) * Atmospherics.N2ODecompositionRate);
```
Comment (Partmedia) [inline Content.Server/Atmos/Reactions/N2ODecompositionReaction.cs:23]: `Atmospherics.N2ODecompositionRate` already goes into the `burnedFuel` calculation, so why multiply it again? Why not just square `Atmospherics.N2ODecompositionRate`?

File: Content.Server/Atmos/Reactions/N2ODecompositionReaction.cs [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+[UsedImplicitly]
+public sealed partial class N2ODecompositionReaction : IGasReactionEffect
+{
+    public ReactionResult React(GasMixture mixture, IGasMixtureHolder? holder, AtmosphereSystem atmosphereSystem, float heatScale)
+    {
+        var cacheN2O = mixture.GetMoles(Gas.NitrousOxide);
+
+        var burnedFuel = cacheN2O * Atmospherics.N2ODecompositionRate * (mixture.Temperature - Atmospherics.N2ODecompositionMinScaleTemp) * (mixture.Temperature - Atmospherics.N2ODecompositionMaxScaleTemp) / Atmospherics.N2ODecompositionScaleDivisor;
+
+        //if (burnedFuel <= 0 || cacheN2O - burnedFuel < 0) return ReactionResult.NoReaction;
```
Comment (Partmedia) [inline Content.Server/Atmos/Reactions/N2ODecompositionReaction.cs:19]: Why is this commented out?

File: Content.Shared/Atmos/Atmospherics.cs [hunk 3]
```diff
@@ -236,6 +236,37 @@ static Atmospherics()
         /// </summary>
         public const float FrezonProductionConversionRate = 50f;
 
+        /// <summary>
+        ///     The minimum temperature N2O can decompose at.
+        /// </summary>
+        public const float N2ODecompositionMinTemp = 1400f;
```
Comment (Partmedia) [inline Content.Shared/Atmos/Atmospherics.cs:242]: This is quite high. Could you explain why you picked this number?

File: Content.Shared/Atmos/Atmospherics.cs [hunk 4]
```diff
@@ -236,6 +236,37 @@ static Atmospherics()
...
+
+        /// <summary>
+        ///     The maximum temperature N2O can decompose at.
+        /// </summary>
+        public const float N2ODecompositionMaxTemp = 100e3f;
```
Comment (Partmedia) [inline Content.Shared/Atmos/Atmospherics.cs:247]: This does not appear to be used.

File: Content.Server/Atmos/Reactions/N2ODecompositionReaction.cs [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+///     Decomposes Nitrous Oxide into Nitrogen and Oxygen.
+/// </summary>
+[UsedImplicitly]
+public sealed partial class N2ODecompositionReaction : IGasReactionEffect
+{
+    public ReactionResult React(GasMixture mixture, IGasMixtureHolder? holder, AtmosphereSystem atmosphereSystem, float heatScale)
+    {
+        var cacheN2O = mixture.GetMoles(Gas.NitrousOxide);
+
+        var burnedFuel = cacheN2O * Atmospherics.N2ODecompositionRate * (mixture.Temperature - Atmospherics.N2ODecompositionMinScaleTemp) * (mixture.Temperature - Atmospherics.N2ODecompositionMaxScaleTemp) / Atmospherics.N2ODecompositionScaleDivisor;
```
Comment (Partmedia) [inline Content.Server/Atmos/Reactions/N2ODecompositionReaction.cs:17]: Why is this a quadratic function? The numbers chosen for `N2ODecompositionMinScaleTemp` and `N2ODecompositionMaxScaleTemp` are also so large that this quadratic is effectively linear in any reasonable temperature range. Could you please talk a bit more about how you came up with this equation and why the numbers are the way they are?

**Review by Partmedia — CHANGES_REQUESTED** [review]
I've taken a look at the code. Could you could please take a look at the review comments in-line?

File: Content.Shared/Atmos/Atmospherics.cs (hunk 4 above)
Comment (whateverusername0) [reply Content.Shared/Atmos/Atmospherics.cs]: true

File: Content.Shared/Atmos/Atmospherics.cs (hunk 3 above)
Comment (whateverusername0) [reply Content.Shared/Atmos/Atmospherics.cs]: i didn't pick this number 🤷‍♂️ it was /tg devs who did

File: Content.Server/Atmos/Reactions/N2ODecompositionReaction.cs (hunk 2 above)
Comment (whateverusername0) [reply Content.Server/Atmos/Reactions/N2ODecompositionReaction.cs]: now it's not

File: Content.Server/Atmos/Reactions/N2ODecompositionReaction.cs (hunk 5 above)
Comment (whateverusername0) [reply Content.Server/Atmos/Reactions/N2ODecompositionReaction.cs]: i didn't came up with it. as i said i violently copied the code from /tg/station 🤷‍♂️

**Comment by whateverusername0** [comment]
yeah, it didn't need to get that complicated

**Comment by Partmedia** [comment]
Thanks for the changes. I've noticed that you've now entirely removed the temperature dependence. I think having a bit of temperature dependence would be good for adding interesting gameplay and would be happy to write one in, but you'll have to give me a day or two.

**Comment by whateverusername0** [comment]
ok

**Review by Partmedia — APPROVED** [review]

**Review by Partmedia — APPROVED** [review]
Actually, we can add the temperature-dependent effects later.

**Pull request merged**